UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,625 @@
+
+        PropertyPtr = FdtGetProperty (Fdt, UPLImageNode, "addr", &TempLen);
+
+        ASSERT (TempLen > 0);
+        if (TempLen > 0) {
+          Data64       = (UINT64 *)(PropertyPtr->Data);
+          StartAddress = Fdt64ToCpu (*Data64);
+          DEBUG ((DEBUG_INFO, "\n         Property(00000000)  entry"));
+          DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+          PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+        }
+      }
+    }
+    // Optional
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb@", AsciiStrLen ("pci-rb@")) == 0) {
+      //
+      // Create PCI Root Bridge Info Hob.
+      //
+      PciRootBridgeInfo = BuildGuidHob (&gUniversalPayloadPciRootBridgeInfoGuid, sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE));
+      ASSERT (PciRootBridgeInfo != NULL);
+      if (PciRootBridgeInfo == NULL) {
+        break;
+      }
+
+      PciRootBridgeInfo->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE);
+      PciRootBridgeInfo->Header.Revision = UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION;
+      PciRootBridgeInfo->Count = 0;
+      PciRootBridgeInfo->ResourceAssigned = FALSE;
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+        if (AsciiStrCmp (TempStr, "ranges") == 0) {
+          PciRootBridgeInfo->Count = 1;
+          //PciRootBridgeInfo->ResourceAssigned = FALSE;
+          PciRootBridgeInfo->RootBridge[0].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;
+          PciRootBridgeInfo->RootBridge[0].Supports             = ROOT_BRIDGE_SUPPORTS_DEFAULT;
+          PciRootBridgeInfo->RootBridge[0].MemAbove4G.Base      = PcdGet64 (PcdPciReservedMemAbove4GBBase);;
+          PciRootBridgeInfo->RootBridge[0].MemAbove4G.Limit     = PcdGet64 (PcdPciReservedMemAbove4GBLimit);
+          PciRootBridgeInfo->RootBridge[0].PMemAbove4G.Base     = PcdGet64 (PcdPciReservedPMemAbove4GBBase);
+          PciRootBridgeInfo->RootBridge[0].PMemAbove4G.Limit    = PcdGet64 (PcdPciReservedPMemAbove4GBLimit);;
+
+          Data64 = (UINT64 *)(PropertyPtr->Data);
+          PciRootBridgeInfo->RootBridge[0].Mem.Base   = Fdt64ToCpu (*Data64);
+          PciRootBridgeInfo->RootBridge[0].Mem.Limit  = Fdt64ToCpu (*(Data64 + 3));
+          PciRootBridgeInfo->RootBridge[0].PMem.Base  = Fdt64ToCpu (*(Data64 + 4));
+          PciRootBridgeInfo->RootBridge[0].PMem.Limit = Fdt64ToCpu (*(Data64 + 7));
+          PciRootBridgeInfo->RootBridge[0].Io.Base    = Fdt64ToCpu (*(Data64 + 8));
+          PciRootBridgeInfo->RootBridge[0].Io.Limit   = Fdt64ToCpu (*(Data64 + 11));
+        }
+        if (AsciiStrCmp (TempStr, "bus-range") == 0) {

```

Comment:
we also support dma-ranges to indicate above 4GB DMA supported or not.
example:
//non-reloc/non-prefetch/memory, child-addr, parent-addr, length
	// indicate rb1 does not support above 4GB DMA
	dma-ranges = <0x82000000 0x0 0x0 0x0 0x0 0x1 0x0>; // 0 ~ 4GB